base,rpc: move transport handling to rpcContext

tbg opened the pull request:
This commit moves the TLS-related functionaliy that has piled up on
`*base.Config` onto a dedicated helper that lives in the `rpc` package.

This isn't perfect yet, but a good step forward - it is now possible to
reason about the certificate names and locations without having to set
up a full `CertificateManager`, and `*base.Config` has lost much of the
mutable state it once had and is much closer to being an actual
"config".

Release note: None

tbg committed:
server: move eager cert load after rpcContext creation

We want to move the certificate handlers off *base.Config, onto
*rpc.Context, so this is a good first step.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/50438)
<!-- Reviewable:end -->

bdarnell reviewed (approved):
> Reviewed 20 of 20 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50438)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell, @nvanbenschoten, and @tbg)

___
*[pkg/security/certificate_manager.go, line 265 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/50438#-MASLDYq5Z9AogXcjn-7:-MASLDYq5Z9AogXcjn-8:b-ryfglu) ([raw file](https://github.com/cockroachdb/cockroach/blob/abd034a07e6951fe1bea4c028f21d3059b4c62a0/pkg/security/certificate_manager.go#L265)):*
> ```Go
> 
> // CACertPath returns the expected file path for the CA certificate.
> func (cd CertsLocator) CACertPath() string {
> ```

s/cd/cl/g


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r6, 20 of 20 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50438#-:-MASRkPz3INvoH8v1zhf:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten and @tbg)

___
*[pkg/rpc/pg.go, line 16 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/50438#-MASRLxb7Ar-nPGMiaNG:-MASRLxb7Ar-nPGMiaNH:bnhelhc) ([raw file](https://github.com/cockroachdb/cockroach/blob/abd034a07e6951fe1bea4c028f21d3059b4c62a0/pkg/rpc/pg.go#L16)):*
> ```Go
> 
> // LoadSecurityOptions extends a url.Values with SSL settings suitable for
> // the given server config. It returns true if and only if the URL
> ```

"It returns true"?


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50438)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @bdarnell and @nvanbenschoten)

___
*[pkg/security/certificate_manager.go, line 265 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/50438#-MASLDYq5Z9AogXcjn-7:-MA_h9OB-1Zfr4vTmfFO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/abd034a07e6951fe1bea4c028f21d3059b4c62a0/pkg/security/certificate_manager.go#L265)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/cd/cl/g
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 19 of 19 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50438#-:-MAaaFc69cep7yWNzzUS:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @bdarnell)



<!-- Sent from Reviewable.io -->

tbg committed:
base,rpc: move transport handling to rpcContext

This commit moves the TLS-related functionaliy that has piled up on
`*base.Config` onto a dedicated helper that lives in the `rpc` package.

This isn't perfect yet, but a good step forward - it is now possible to
reason about the certificate names and locations without having to set
up a full `CertificateManager`, and `*base.Config` has lost much of the
mutable state it once had and is much closer to being an actual
"config".

Release note: None

tbg commented:
bors r=nvanbenschoten
TFTRs!
